File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/mobs/Skeleton.java

package com.watabou.pixeldungeon.actors.mobs;

import com.nyrds.pixeldungeon.items.Treasury;
import com.nyrds.pixeldungeon.levels.NecroBossLevel;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.pixeldungeon.mobs.necropolis.UndeadMob;
import com.nyrds.platform.audio.Sample;
import com.nyrds.platform.util.StringsManager;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.ResultDescriptions;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.utils.GLog;
import com.watabou.pixeldungeon.utils.Utils;

import org.jetbrains.annotations.NotNull;

public class Skeleton extends UndeadMob {

	public Skeleton() {
		hp(ht(25));
		baseDefenseSkill = 9;
		baseAttackSkill  = 12;
		dmgMin = 3;
		dmgMax = 8;
		dr = 5;
		
		expForKill = 5;
		maxLvl = 10;

		loot(getLoot(), 0.2f);
	}

	@Override
	public Object getLoot() {
		if (!(level() instanceof NecroBossLevel)) {
			return Treasury.getLevelTreasury().worstOf(Treasury.Category.WEAPON,3 );
		}
		return null;
	}

	@Override
	public void die(@NotNull NamedEntityKind cause) {
		
		super.die( cause );
		
		boolean heroKilled = false;
		for (int i=0; i < Level.NEIGHBOURS4.length; i++) {
			Char ch = findChar( getPos() + Level.NEIGHBOURS4[i] );
			if (ch != null && ch.isAlive()) {
				int damage = Math.max( 0,  damageRoll() - ch.defenceRoll(this) / 2 );
				ch.damage( damage, this );
				if (ch == Dungeon.hero && !ch.isAlive()) {
					heroKilled = true;
				}
			}
		}
		
		if (CharUtils.isVisible(this)) {
			Sample.INSTANCE.play( Assets.SND_BONES );
		}
		
		if (heroKilled) {
			Dungeon.fail( Utils.format( ResultDescriptions.getDescription(ResultDescriptions.Reason.MOB), Utils.indefinite( getName() ), Dungeon.depth ) );
            GLog.n(StringsManager.getVar(R.string.Skeleton_Killed));
		}
	}
}

File: scripts/mobs/Skeleton.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

-- java Skeleton (batch 8): carries the worst of 3 random weapons (20%),
-- detonates on death - bone-blast damage to all 4 neighbours.
return mob.init{
    stats = function(self)
        -- ctor-only roll, persisted (stats re-runs on every restore)
        local data = mob.restoreData(self)
        if data.rolled then
            return
        end
        if not self:getBelongings():isBackpackEmpty() then
            -- pre-migration saves already carry the weapon their java rolled
            data.rolled = true
            return
        end
        data.rolled = true

        -- java getLoot(): no treasury loot on the Lich's arena
        -- (ctor runs during level creation, Dungeon.level may be nil -
        -- java's `level() instanceof NecroBossLevel` was null-safe too)
        local lvl = RPD.Dungeon.level
        if lvl ~= nil and lvl:levelKind() == "NecroBossLevel" then
            return
        end

        -- java loot() hero-level gate
        local hero = RPD.Dungeon.hero
        if not self:isBoss() and hero:lvl() > self:getMaxLvl() + 2 + self:lvl() then
            return
        end

        if math.random(5) == 1 then
            self:collect(RPD.Treasury:getLevelTreasury():worstOf("WEAPON", 3))
        end
    end,

    die = function(self, cause)
        -- bone-blast: damage all 4 neighbours
        local level = RPD.Dungeon.level
        local pos = self:getPos()
        local w = level:getWidth()
        local heroKilled = false
        for _, off in ipairs({-w, 1, w, -1}) do
            local ch = RPD.Actor:findChar(pos + off)
            if ch ~= nil and ch:isAlive() then
                local dmg = math.max(0, self:damageRoll() - ch:defenceRoll(self) / 2)
                ch:damage(dmg, self)
                if ch:getEntityKind() == "Hero" and not ch:isAlive() then
                    heroKilled = true
                end
            end
        end

        if RPD.CharUtils:isVisible(self) then
            RPD.playSound("snd_bones")
        end

        if heroKilled then
            -- java Skeleton.die set the death report explicitly (on top of the
            -- generic Doom path, which composes the same MOB wording)
            RPD.Dungeon:fail(RPD.JavaUtils:format(
                    RPD.ResultDescriptions:getDescription(RPD.ResultReason.MOB),
                    { RPD.JavaUtils:indefinite(self:getName()), RPD.Dungeon.depth }))
            RPD.glogn(RPD.textById("Skeleton_Killed"))
        end
    end
}
